src/output/fmt.rs

```diff
@@ -1,187 +1,120 @@
+
+use std::fmt::{self, Debug, Formatter};
+use std::io::{Result, Write};
 
 #[cfg(feature = "color")]
-pub(crate) fn is_a_tty(stderr: bool) -> bool {
+fn is_a_tty(stderr: bool) -> bool {
     debugln!("is_a_tty: stderr={:?}", stderr);
+
     let stream = if stderr {
         atty::Stream::Stderr
     } else {
         atty::Stream::Stdout
     };
+
     atty::is(stream)
 }
 
 #[cfg(not(feature = "color"))]
-pub(crate) fn is_a_tty(_: bool) -> bool {
+fn is_a_tty(_: bool) -> bool {
     debugln!("is_a_tty;");
     false
 }
 
-pub(crate) fn is_term_dumb() -> bool {
-    env::var("TERM").ok() == Some(String::from("dumb"))
-}
-
-pub(crate) struct ColorizerOption {
-    pub(crate) use_stderr: bool,
-    pub(crate) when: ColorWhen,
+pub struct Colorizer {
+    writer: BufferWriter,
+    buffer: Buffer,
 }
 
-pub(crate) struct Colorizer {
-    when: ColorWhen,
-}
+impl Debug for Colorizer {
+    #[cfg(feature = "color")]
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", String::from_utf8_lossy(self.buffer.as_slice()))
+    }
 
-macro_rules! color {
-    ($_self:ident, $c:ident, $m:expr) => {
-        match $_self.when {
-            ColorWhen::Auto => Format::$c($m),
-            ColorWhen::Always => Format::$c($m),
-            ColorWhen::Never => Format::None($m),
-        }
-    };
+    #[cfg(not(feature = "color"))]
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", String::from_utf8_lossy(&self.buffer))
+    }
 }
 
 impl Colorizer {
-    pub(crate) fn new(option: &ColorizerOption) -> Colorizer {
-        let is_a_tty = is_a_tty(option.use_stderr);
-        let is_term_dumb = is_term_dumb();
-        Colorizer {
-            when: if is_a_tty && !is_term_dumb {
-                option.when
-            } else {
-                ColorWhen::Never
-            },
-        }
+    pub fn new(use_stderr: bool, when: ColorChoice) -> Self {
+        let checked_when = if is_a_tty(use_stderr) {
+            when
+        } else {
+            ColorChoice::Never
+        };
+
+        let writer = if use_stderr {
+            BufferWriter::stderr(checked_when)
+        } else {
+            BufferWriter::stdout(checked_when)
+        };
+
+        let buffer = writer.buffer();
+
+        Self { writer, buffer }
     }
 
-    pub(crate) fn good<T>(&self, msg: T) -> Format<T>
-    where
-        T: fmt::Display + AsRef<str>,
-    {
-        debugln!("Colorizer::good;");
-        color!(self, Good, msg)
+    pub fn print(&mut self) -> Result<()> {
+        self.writer.print(&self.buffer)
     }
 
-    pub(crate) fn warning<T>(&self, msg: T) -> Format<T>
-    where
-        T: fmt::Display + AsRef<str>,
-    {
-        debugln!("Colorizer::warning;");
-        color!(self, Warning, msg)
+    #[cfg(feature = "color")]
+    pub fn good(&mut self, msg: &str) -> Result<()> {
+        self.buffer
+            .set_color(ColorSpec::new().set_fg(Some(Color::Green)))?;
```

Comment:
This is what I was talking about: if `set_color` returns `Err`, it will be propagated up to `write!`, which will panic. But why do we need to panic in the first place? If coloring has failed. this likely means that color support detection misfired somehow. The most sensible way to proceed is to ignore the error and go in "no color" mode, I think.

Reply:
That would be something for termcolor to handle. This Err is currently propogated up and an error is written to `io` saying issues with `stderr` or `stdout` which I think is reasonable in the meanwhile without messing up our performance.

Reply:
That's definitely not for termcolor to handle. It's job is to attempt to detect, attempt to set the color if detected, propagate the error to the developer, if any. It's developer's decision what to do: ignore it, print it, crash the program. As a user, I would expect this kind of errors to be silently ignored, but it's not so important to block the fix on it. 

Maybe I'm just being paranoid and nobody ever will come with such bug report? Probably.